rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,281 @@
+   are frequently subject to this type of pattern). Using insertion timestamps to resolve which
+   version of a key is more recent may be preferable to using the clock at the time of insertion.
+3. **Historical Reads / MVCC**: while SlateDB provides snapshot capabilities to ensure reads are
+   consistent and repeatable, it requires a snapshot to have been taken at the desired time. 
+   increasing the number of snapshots adds significant storage overhead. Instead, it is possible
+   to trade read-latency for space amplification by maintaining multiple versions of keys at
+   different timestamps to allow for similar "historical" read operations.
+
+## Goals
+
+- Define the API for specifying timestamps
+- Define the changes to the on-disk format of stored rows
+- Define the default semantics (and enumerate configurations) for out-of-order insertions
+- Outline mechanism to support Time-To-Live (TTL) via Compaction Filters
+
+## Out of Scope
+
+While the following are out of scope for the initial implementation, the design should
+not preclude:
+
+- Outlining compaction strategies that are optimized for TTL (such as Cassandra's
+  [TWCS](https://cassandra.apache.org/doc/stable/cassandra/operating/compaction/twcs.html)). This
+  will require additional metadata per-SST and a compaction strategy to match, but is otherwise
+  possible with what's outlined in the scope of this design.
+- Outlining compaction strategies that target efficient deletes (such as outlined
+  in [Lethe](https://disc-projects.bu.edu/lethe/))
+- Ensuring deletions are resolved within a time bound (a TTL does not ensure the data is deleted
+  from the object store within the specified time)
+- A public API for compaction filters (see https://github.com/slatedb/slatedb/issues/225) - this RFC
+  will implement TTL using a compaction filter but will not yet expose a public API for plugging in
+  custom compaction filters. Instead, the implementation should ensure that such an extension is
+  easy.
+
+## Public API
+
+### Configuration
+
+We will make the following changes to `DbOptions`:
+- introduce a configuration for a custom clock, and will default to using the system clock
+- introduce a configuration for a default TTL, which defaults to -1
+
+```rust
+/// defines the clock that SlateDB will use during this session
+pub trait Clock {
+    fn now(&self) -> i64;
+}
+
+/// contains the default implementation of the Clock, and will return the system time
+pub struct SystemClock;
+
+// ...
+
+pub struct DbOptions {
+    // ...
+
+    /// the Clock to use for insertion timestamps
+    ///
+    /// default: the default clock uses the local system time on the machine
+    clock: Option<Arc<dyn Clock>>,
+
+    /// the default time-to-live (TTL) for insertions (note that re-inserting a key
+    /// with any value will update the TTL to use the default_ttl)
+    ///
+    /// default: no TTL (insertions will remain until deleted)
+    default_ttl: Option<Duration>
```

Comment:
this will be particularly useful (the knowledge that every row has a TTL) if we want to encode the min_ts/max_ts in an SST. That way we can drop the entire SST without looking at it at all if there's a default ttl. 

that isn't covered in this proposal since it requires a bit more work (i.e. handling the interaction between the default ttl and the per-row ttl). the way it would work as proposed would be that you could only drop the entire SST if both there's a default ttl _and_ the `MetaFeature` for `TimeToLive` is not present. there may be a way we could expand the configuration to have a "max ttl" instead of just a "default ttl"...

Anyway, it's a bit complicated so I left it out of this RFC for now.